### ISSUE TITLE
SQL Invalid parameter error because of ValueBinder

### ISSUE THREAD

**Issue by intelligide** [cakephp/cakephp#8682]
This is a (multiple allowed):
* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.2.7
* Platform and Target: Windows, PHP 7.0.0, MariaDB 10

### What I did

```php
$results = TableRegistry::get("UserInformations")->find()
            ->where(function(QueryExpression $exp) use($text){
                return $exp->add("MATCH(name,hei_id) AGAINST(:q)");
            })
            ->bind("q",$text)
            ->page($optn["page"], $optn["count"])
            ->toList();
```

### Expected Behavior
Just sending request without parameter error

### Actual Behavior
I got an SQL error because ValueBinder doesn't bind value
```
 Invalid parameter number: parameter was not defined
```

### Fix
Line 54 of Database/ValueBinder is wrong:
```
'placeholder' => is_int($param) ? $param : substr($param, 1)
```
should be
```
'placeholder' => is_int($param) ? substr($param, 1): $param
```

Thanks

**Comment by markstory** [comment]
Why does `substr()` on an int fix the problem?

**Comment by intelligide** [comment]
```substr()``` doesn't fix the problem, but ternary expressions are just inverted

**Comment by ionas** [comment]
So you are saying `'placeholder' => (string) $param` fixes your issue?
I think @markstory is not saying that there could not be a bug, but that your suggested fix is strange in it self.

**Comment by thinkingmedia** [comment]
If the `$param` is int the binding is referred to by index. If the `$param` is string the binding is referred to by name, but the string is prefixed with `:` so `substr($param,1)` drops the prefix.

Changing the ternary is not fixing the issue. This method should be doing validation on the `$param` to ensure that it is formatted as `:name` when used as a string.

```
Invalid parameter number: parameter was not defined
```

The above error means that a numeric parameter identifier was used in the SQL and a value was not bound to that index when the statement was prepared. 

```
'placeholder' => is_int($param) ? substr($param, 1): $param
```

When you do this to a parameter value of `"q"` as in your example it results in `placeholder` being an empty string. It is not fixing the problem. It just just assigning empty to placeholder.

I suspect that when placeholder is empty that the datasource handles that as a `0` numeric index. Which is what the prepared statement was looking for.

So to summarize. 

- error message means statement has numeric parameter that is not bound
- changing `is_int` logic resulted in empty placeholder
- empty string equals `0` integer
- the missing numeric parameter was most likely `0` and so the error went away

**Comment by markstory** [comment]
@intelligide Ok, but inverting the ternary will have other adverse effects. I'll see if I can reproduce this issue and come up with a safer fix.

**Comment by markstory** [comment]
The issue here is that your bind() parameter name needs to include the `:`. In your case it should be `:q`. This requirement isn't clear from the API docs, and will be corrected.

**Comment by markstory** [comment]
Closing as everything works when you use `:q` as your placeholder.
